sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp

```diff
@@ -376,42 +379,82 @@ template <class PassClass> static bool runModulePass(llvm::Module &M) {
+
+  // TODO: The optlevel passed to `sycl-post-link` is determined by
+  //       `getSYCLPostLinkOptimizationLevel` in
+  //       `clang/lib/Driver/ToolChains/Clang.cpp`.
+  const bool PerformOpts = true;
+
+  // Propagate ESIMD attribute to wrapper functions to prevent spurious splits
+  // and kernel link errors.
+  runModulePass<SYCLFixupESIMDKernelWrapperMDPass>(*Module);
+
+  assert(!Module->getGlobalVariable("llvm.used") &&
+         !Module->getGlobalVariable("llvm.compiler.used"));
   // Otherwise: Port over the `removeSYCLKernelsConstRefArray` and
   // `removeDeviceGlobalFromCompilerUsed` methods.
 
-  assert(!isModuleUsingAsan(Module));
+  assert(!isModuleUsingAsan(*Module));
   // Otherwise: Need to instrument each image scope device globals if the module
   // has been instrumented by sanitizer pass.
 
   // Transform Joint Matrix builtin calls to align them with SPIR-V friendly
   // LLVM IR specification.
-  runModulePass<SYCLJointMatrixTransformPass>(Module);
+  runModulePass<SYCLJointMatrixTransformPass>(*Module);
+
+  // Do invoke_simd processing before splitting because this:
+  // - saves processing time (the pass is run once, even though on larger IR)
+  // - doing it before SYCL/ESIMD splitting is required for correctness
+  if (runModulePass<SYCLLowerInvokeSimdPass>(*Module)) {
+    return createStringError("`invoke_simd` calls detected");
+  }
 
   // TODO: Implement actual device code splitting. We're just using the splitter
   //       to obtain additional information about the module for now.
-  // TODO: EmitOnlyKernelsAsEntryPoints is controlled by
-  //       `shouldEmitOnlyKernelsAsEntryPoints` in
-  //       `clang/lib/Driver/ToolChains/Clang.cpp`.
+
   std::unique_ptr<ModuleSplitterBase> Splitter = getDeviceCodeSplitter(
-      ModuleDesc{std::unique_ptr<llvm::Module>{&Module}}, SPLIT_NONE,
-      /*IROutputOnly=*/false,
-      /*EmitOnlyKernelsAsEntryPoints=*/true);
-  assert(Splitter->remainingSplits() == 1);
+      ModuleDesc{std::move(Module)}, SplitMode,
+      /*IROutputOnly=*/false, EmitOnlyKernelsAsEntryPoints);
+  assert(Splitter->hasMoreSplits());
+  if (Splitter->remainingSplits() > 1) {
+    return createStringError("Device code requires splitting");
+  }
 
   // TODO: Call `verifyNoCrossModuleDeviceGlobalUsage` if device globals shall
   //       be processed.
 
-  assert(Splitter->hasMoreSplits());
   ModuleDesc MDesc = Splitter->nextSplit();
-  assert(&Module == &MDesc.getModule());
+
+  // TODO: Call `MDesc.fixupLinkageOfDirectInvokeSimdTargets()` when
+  //       `invoke_simd` is supported.
+
+  SmallVector<ModuleDesc, 2> ESIMDSplits =
+      splitByESIMD(std::move(MDesc), EmitOnlyKernelsAsEntryPoints);
+  assert(!ESIMDSplits.empty());
+  if (ESIMDSplits.size() > 1) {
+    return createStringError("Mixing SYCL and ESIMD code is unsupported");
+  }
+  MDesc = std::move(ESIMDSplits.front());
+
+  if (MDesc.isESIMD()) {
+    // TODO: We're assuming ESIMD lowering is not deactivated (why would it?).
```

Comment:
The driver would deactivate it (i.e. not pass `-lower-esimd` to `sycl-post-link`) in the IR-only output mode, which is only relevant for spec constant processing IIUC. Otherwise it doesn't seem to be influenced by a user-option. I updated the comment to reflect this.